### DISCUSSION TITLE
What is the best library for mocking dependencies compatible with AVA?

### DISCUSSION THREAD

**Discussion by arturtamborski** [avajs/ava#2727] [Q&A] [answered]
Hi,

I'm looking for the best (or realistically speaking, any) library that can be used for mocking external deps from node modules so that I can write better tests.

Any suggestions as to what should I use? I heard about `rewiremock`, but it's relatively to use (compared to `jest`).

**Comment by novemberborn** [comment, marked as answer]
Anything that works with Node.js is compatible with AVA.

I don't find myself needing to mock dependencies very often, and I think it's something best avoided, but I can't speak to your code.

**Reply by arturtamborski** [reply]
You're right about the uncommon need for mocking, only after I wrote the original question I stumbled upon this wonderful document advising to _not_ mock what's not owned (as in, third party deps).
https://github.com/testdouble/contributing-tests/wiki/Don't-mock-what-you-don't-own

Thanks for the answer, I'm marking this one as resolved (as you said, anything that works is compatible).

**Reply by phishy** [reply]
@novemberborn I'm curious about your statement. If you're not mocking network, file, and shell calls.. are you writing everything with DI?

**Reply by novemberborn** [reply]
@phishy at the time I was mostly testing services, and took an integration testing kind of approach.

 When those services call _other_ services, we'd spin up a mock of that other service and configure it through environment variables. The service under test would be launched in a child process. We'd then fire requests at it from the test suite and assert return values or database state.

Rather than mocking actual dependencies like cloud SDKs we'd have a wrapper module with informal dependency injection (an exported function that, under test, would allow the dependency to be swapped out by some other code that could be observed by the tests).

There's a bit of manual setup and repetition of the pattern across services, but it's also fairly easy to see what's going on.

**Reply by bn-l** [reply]
> Rather than mocking actual dependencies like cloud SDKs we'd have a wrapper module with informal dependency injection (an exported function that, under test, would allow the dependency to be swapped out by some other code that could be observed by the tests).

Sorry about the necro but isn't that a mock but with a wrapper in front of it? (genuinely asking)

**Reply by novemberborn** [reply]
Sure. The key thing is that the wrapper lets you inject the mock. We're not mocking anything coming from `node_modules`.
